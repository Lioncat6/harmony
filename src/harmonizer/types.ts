import type { PartialDate } from '../utils/date.ts';

export type HarmonyRelease = {
	title: string;
	artists: ArtistCredit;
	gtin: GTIN;
	externalLinks: URL[];
	media: HarmonyMedium[];
	releaseDate: PartialDate;
	labels?: Label[];
	images?: Artwork[];
};

export type HarmonyMedium = {
	title?: string;
	number?: number;
	tracklist: HarmonyTrack[];
};

export type HarmonyTrack = {
	title: string;
	artists?: ArtistCredit;
	number: number | string;
	/** Track duration in milliseconds. */
	duration: number;
	isrc?: string;
};

export type ArtistCreditName = {
	name: string;
	creditedName?: string;
	externalLink?: URL;
	joinPhrase?: string;
};

type ArtistCredit = ArtistCreditName[];

export type Label = {
	name: string;
	externalLink?: URL;
	catalogNumber?: string;
};

export type Artwork = {
	url: URL;
	types?: ArtworkType[];
	comment?: string;
};

type ArtworkType = 'front' | 'back' | 'thumbnail';

/** Global Trade Item Number with 8 (EAN-8), 12 (UPC), 13 (EAN-13) or 14 digits. */
export type GTIN = number | string;

export type ReleaseOptions = Partial<{
	withSeparateMedia: boolean;
	withISRC: boolean;
	withCountryAvailability: boolean;
}>;

export type ProviderName = string;

/** Mapping from the provider's name to the release returned by that provider. */
export type ProviderReleaseMapping = Record<ProviderName, HarmonyRelease | undefined>;

export type ReleaseProperty = keyof HarmonyRelease;

/** Mapping from release properties to lists of preferred providers for these properties */
export type ProviderPreferences = Partial<Record<ReleaseProperty, ProviderName[]>>;
