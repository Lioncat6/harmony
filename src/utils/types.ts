export type MaybeArray<T> = T | T[];

export type MaybePromise<T> = T | Promise<T>;

export type FormDataRecord = Record<string, MaybeArray<string>>;

export type Key = string | number | symbol;

interface NestedRecord {
	[key: Key]: NestedRecord | MaybeArray<Key>;
}
