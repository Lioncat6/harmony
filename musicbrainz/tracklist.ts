import type { ArtistCreditName, HarmonyTrack } from '@/harmonizer/types.ts';
import { formatDuration } from '@/utils/time.ts';

export interface TracklistFormat {
	useNumber: boolean;
	useArtist: boolean;
	artistDelimiter: string;
}

/** Formats a tracklist which is compatible with the MusicBrainz track parser format. */
export function formatTracklist(tracks: HarmonyTrack[], format: Partial<TracklistFormat> = {}): string {
	return tracks.map((track, index) => {
		let line = '';
		if (format.useNumber !== false) {
			line = `${track.number ?? index + 1}. `;
		}
		line += track.title;
		if (track.artists && format.useArtist !== false) {
			line += ` ${format.artistDelimiter ?? '-'} ${joinArtistCredit(track.artists)}`;
		}
		if (track.length) {
			line += ` (${formatDuration(track.length, { showMs: false })})`;
		}
		return line;
	}).join('\n');
}

/** Combines the given artist credits into a string. */
function joinArtistCredit(
	credits: ArtistCreditName[],
): string {
	const lastIndex = credits.length - 1;
	return credits.flatMap((credit, index) => {
		const defaultJoinPhrase = (index !== lastIndex) ? (index === lastIndex - 1 ? ' & ' : ', ') : undefined;
		return [
			credit.creditedName ?? credit.name,
			credit.joinPhrase ?? defaultJoinPhrase,
		];
	}).join('');
}
