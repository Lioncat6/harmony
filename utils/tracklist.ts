import type { HarmonyRelease } from '@/harmonizer/types.ts';
import { plural } from '@/utils/plural.ts';

/** Creates a per medium track count summary for the given release. */
export function trackCountSummary(release: HarmonyRelease): string {
	const trackCounts = release.media.map((medium) => medium.tracklist.length);
	const totalTrackCount = trackCounts.reduce((sum, current) => sum + current, 0);

	return `${trackCounts.join('+')} ${plural(totalTrackCount, 'track')}`;
}
