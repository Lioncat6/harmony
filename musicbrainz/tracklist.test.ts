import { describe, it } from '@std/testing/bdd';
import { HarmonyTrack } from '@/harmonizer/types.ts';
import { assertSnapshot } from '@std/testing/snapshot';
import { formatTracklist } from '@/musicbrainz/tracklist.ts';

describe('formatTracklist', () => {
	const dummyTracks: HarmonyTrack[] = [{
		title: 'Track only has a title',
	}, {
		title: 'Track with rounded length',
		length: 123654,
	}, {
		title: 'Track with a single artist',
		artists: [{
			name: 'Artist',
		}],
		number: 4,
		length: 231456,
	}, {
		title: 'Track with multiple artists',
		artists: [{
			name: 'Artist 1',
		}, {
			name: 'Artist 2',
		}, {
			name: 'Artist 3',
		}],
		number: 'B7',
		length: 42000,
	}, {
		title: 'Track with artist credit names and join phrase',
		artists: [{
			name: 'Artist 1',
			creditedName: 'Credit 1',
			joinPhrase: ' feat. ',
		}, {
			name: 'Artist 2',
		}, {
			name: 'Artist 3',
			creditedName: 'Credit 3',
		}],
		number: '123',
		length: 654321,
	}];

	it('formats tracks with default options', async (ctx) => {
		await assertSnapshot(ctx, formatTracklist(dummyTracks));
	});
});
