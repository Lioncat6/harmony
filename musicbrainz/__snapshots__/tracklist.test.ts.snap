export const snapshot = {};

snapshot[`formatTracklist > formats tracks with default options 1`] = `
"1. Track only has a title
2. Track with rounded length (2:04)
4. Track with a single artist - Artist (3:51)
B7. Track with multiple artists - Artist 1, Artist 2 & Artist 3 (0:42)
123. Track with artist credit names and join phrase - Credit 1 feat. Artist 2 & Credit 3 (10:54)"
`;
